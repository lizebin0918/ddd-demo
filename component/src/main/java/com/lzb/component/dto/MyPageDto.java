package com.lzb.component.dto;

import java.util.List;
import java.util.function.Function;

import lombok.AllArgsConstructor;
import lombok.Getter;
import one.util.streamex.StreamEx;

/**
 * <br/>
 * Created on : 2023-09-22 23:07
 * @author lizebin
 */
@Getter
@AllArgsConstructor
public class MyPageDto<T> {

    private final long pages;
    private final long pageSize;
    private final long total;
    private final List<T> rows;

    private <E> MyPageDto(long pages, long pageSize, long total, List<E> rows, Function<E, T> convertor) {
        this.pages = pages;
        this.pageSize = pageSize;
        this.total = total;
        this.rows = StreamEx.of(rows).map(convertor).toList();
    }

    public static <E, T> MyPageDto<T> of(long pages,
            long pageSize,
            long total,
            List<E> rows,
            Function<E, T> convertor) {
        return new MyPageDto<>(pages, pageSize, total, rows, convertor);
    }
}
