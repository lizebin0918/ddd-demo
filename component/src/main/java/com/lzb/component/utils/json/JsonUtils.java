package com.lzb.component.utils.json;

import java.math.BigDecimal;
import java.time.ZoneId;
import java.util.Locale;
import java.util.Objects;
import java.util.TimeZone;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.json.JsonReadFeature;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.databind.json.JsonMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.fasterxml.jackson.databind.ser.std.ToStringSerializer;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;
import io.vavr.jackson.datatype.VavrModule;
import lombok.experimental.UtilityClass;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.Nullable;

@UtilityClass
@Slf4j
public class JsonUtils {

    public static final JsonMapper OBJECT_MAPPER = JsonMapper.builder()
            .configure(JsonReadFeature.ALLOW_UNESCAPED_CONTROL_CHARS, true)
            .configure(JsonReadFeature.ALLOW_BACKSLASH_ESCAPING_ANY_CHARACTER, true)
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
            .configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true)
            .configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
            // 枚举写成字符串
            .configure(SerializationFeature.WRITE_ENUMS_USING_TO_STRING, true)
            // 时间序列化
            .configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
            // 时区
            .defaultLocale(Locale.CHINA)
            .defaultTimeZone(TimeZone.getTimeZone(ZoneId.systemDefault()))
            .addModule(new JavaTimeModule())
            .addModule(new Jdk8Module())
            // 通过构造函数声明@JsonCreator反序列化，但是要在maven引入compile插件，加参数:-parameters
            .addModule(new ParameterNamesModule(JsonCreator.Mode.PROPERTIES))
            .addModule(new SimpleModule().addSerializer(BigDecimal.class, new ToStringSerializer()))
            .addModule(new VavrModule())
            .build();

   static {
        // 输入非空字段
        OBJECT_MAPPER.setSerializationInclusion(JsonInclude.Include.NON_NULL);
        // 只显示有的字段
        OBJECT_MAPPER.setVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE);
        OBJECT_MAPPER.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
    }

    /**
     * 将对象序列化成json字符串
     *
     * @param value javaBean
     * @param <T>   T 泛型标记
     * @return jsonString json字符串
     */
    @Nullable
    public static <T> String toJSONString(T value) {
        try {
            return OBJECT_MAPPER.writeValueAsString(value);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
        return null;
    }



    /**
     * 将json反序列化成对象
     *
     * @param content   content
     * @param valueType class
     * @param <T>       T 泛型标记
     * @return Bean
     */
    public static <T> T json2JavaBean(String content, Class<T> valueType) {
        if (Objects.isNull(content)) {
            return null;
        }
        try {
            return OBJECT_MAPPER.readValue(content, valueType);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            throw new RuntimeException(e);
        }
    }

}