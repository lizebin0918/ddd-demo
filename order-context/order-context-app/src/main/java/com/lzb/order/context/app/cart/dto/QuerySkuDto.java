package com.lzb.order.context.app.cart.dto;

import lombok.Data;

/**
 * sku查询<br/>
 * Created on : 2023-12-27 19:45
 * @author lizebin
 */
@Data
public class QuerySkuDto {

    private int pageIndex;

    private int pageSize;

}
