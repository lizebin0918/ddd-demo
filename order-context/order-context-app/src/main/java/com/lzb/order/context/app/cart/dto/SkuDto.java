package com.lzb.order.context.app.cart.dto;

import lombok.Builder;

@Builder
public record SkuDto(Integer skuId, String title, String color, String size) {

}
