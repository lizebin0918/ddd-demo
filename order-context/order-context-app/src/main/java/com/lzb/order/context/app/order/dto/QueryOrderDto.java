package com.lzb.order.context.app.order.dto;

import java.util.Set;

/**
 * 订单查询<br/>
 * Created on : 2023-09-06 13:46
 * @author lizebin
 */
public record QueryOrderDto(
        Set<String> orderIds,
        String email,
        int pageIndex,
        int pageSize) {

}
