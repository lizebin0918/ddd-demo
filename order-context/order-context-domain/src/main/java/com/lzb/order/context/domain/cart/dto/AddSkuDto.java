package com.lzb.order.context.domain.cart.dto;

import com.lzb.order.context.domain.cart.aggregation.CartSku;
import java.util.List;

import com.lzb.order.context.domain.core.valobj.SkuCount;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.Positive;
import lombok.Builder;

/**
 * 批量添加sku<br/>
 * Created on : 2023-12-27 11:39
 *
 * @author lizebin
 */
public record AddSkuDto(
    @NotEmpty(message = "The sku code collection cannot be empty")
    List<Sku> skus) {

    @Builder
    public record Sku(
        @NotBlank(message = "The sku code cannot be empty")
        Integer skuId,

        @Positive(message = "Quantity must be greater than 0")
        int count) {

    }

    public List<SkuCount> skuCounts() {
        return skus.stream()
            .map(sku -> new SkuCount(sku.skuId(), sku.count()))
            .toList();
    }

    public List<CartSku> toCartSkus() {
        return skus.stream()
            .map(this::toCartSku)
            .toList();
    }

    private CartSku toCartSku(Sku sku) {
        return CartSku.builder().skuId(sku.skuId()).count(sku.count()).build();
    }

}
