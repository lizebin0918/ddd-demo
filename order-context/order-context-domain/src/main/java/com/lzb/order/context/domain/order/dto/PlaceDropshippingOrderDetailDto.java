package com.lzb.order.context.domain.order.dto;

import java.math.BigDecimal;

import jakarta.validation.constraints.NotBlank;
import lombok.EqualsAndHashCode;
import lombok.Value;

@Value
@EqualsAndHashCode(callSuper = true)
public class PlaceDropshippingOrderDetailDto extends PlaceOrderDetailDto {

    @NotBlank(message = "unitPrice cannot be empty")
    String unitPrice;

    /**
     * @param skuId
     * @param quantity
     * @param count
     * @param unitPrice
     */
    public PlaceDropshippingOrderDetailDto(Integer skuId, Integer quantity, Integer count, String unitPrice) {
        super(skuId, new BigDecimal(quantity), count);
        this.unitPrice = unitPrice;
    }

}

// 另一种写法
/*
@Getter
@EqualsAndHashCode(callSuper = true)
@Jacksonized
@SuperBuilder
public class PlaceDropshippingOrderDetailDto extends PlaceOrderDetailDto {

    @NotNull
    @NotBlank(message = "unitPrice cannot be empty")
    private final String unitPrice;

}*/
