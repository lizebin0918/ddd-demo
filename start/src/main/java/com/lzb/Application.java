package com.lzb;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.autoconfigure.data.redis.RedisRepositoriesAutoConfiguration;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.scheduling.annotation.EnableAsync;

// 启动报错：Multiple Spring Data modules found, entering strict repository configuration
@SpringBootApplication(exclude = { RedisRepositoriesAutoConfiguration.class})
// AopContext.currentProxy() 启用自身调用aop
@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

}
