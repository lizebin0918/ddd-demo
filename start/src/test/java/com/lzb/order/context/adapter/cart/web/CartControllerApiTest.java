package com.lzb.order.context.adapter.cart.web;

import com.lzb.BaseApiTest;
import com.lzb.component.utils.json.JsonUtils;
import com.lzb.order.context.domain.cart.dto.AddSkuDto;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import io.restassured.internal.http.Status;
import java.util.Arrays;
import org.approvaltests.Approvals;
import org.approvaltests.approvers.ApprovalApprover;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static io.restassured.RestAssured.given;

class CartControllerApiTest extends BaseApiTest {

    @Test
    @DisplayName("查询购物车商品列表")
    void should_list_cart_sku() {
        given()
            .when()
            .get("/cart/sku")
            .then()
            .statusCode(200);
    }

    @Test
    @DisplayName("添加购物车商品")
    void should_add_skus_to_cart() {

        AddSkuDto addSkuDto = new AddSkuDto(
            Arrays.asList(AddSkuDto.Sku.builder().skuId(1).count(1).build()));

        given()
            .body(addSkuDto)
            .when()
            .post("/cart/sku")
            .then()
            .statusCode(204);

        String print = given()
            .when()
            .get("/cart/sku")
            .getBody()
            .print();

        Approvals.verify(print);
    }

}