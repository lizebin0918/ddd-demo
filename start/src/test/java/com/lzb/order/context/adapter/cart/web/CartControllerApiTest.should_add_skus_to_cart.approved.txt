{"pages":0,"pageSize":0,"total":1,"rows":[{"skuId":1,"title":null,"color":null,"size":null,"count":1}]}